internal/terraform/graph_builder.go

```diff
@@ -58,6 +61,13 @@ func (b *BasicGraphBuilder) Build(path addrs.ModuleInstance) (*Graph, tfdiags.Di
 		}
 	}
 
+	// Return early if the graph validation is skipped
+	// This behavior is currently only used by the graph command
+	// which only want to display the dot representation of the graph
+	if b.SkipGraphValidation {
+		return g, diags
+	}
+
 	if err := g.Validate(); err != nil {
```

Comment:
This isn't the only `Validate` call, it needs to be called earlier too (this actually should never fail right now). You can't run `TransitiveReduction` if there are cycles, so we call `Validate` at least once when building the graph.

Reply:
There are 3 `Validate` calls. This one, the one in `TransitiveReductionTransfomer`, and one that is called on a subgraph during the graph walk. 

This validation skip is only about building the graph, which is already done before a walk, so I think the latter matters less here. (maybe it would help for it to be called `SkipGraphBuildValidation`
Luckily, [the one in the transitive reduction transformer](https://github.com/hashicorp/terraform/blob/53c34ff49cfbc1f70d7cdd3dca8040551c53737a/internal/terraform/transform_transitive_reduction.go#L11-L23) does not return an error, it is simply a gate to prevent running `TransitiveReduction` on a cyclic graph. 


However, I could try a different approach which would put the skipping step in the `Validate` method itself, so that every where it is called gets that behaviour

Reply:
What do you think of this other approach? @jbardin 
In https://github.com/hashicorp/terraform/pull/36186/commits/46750201ac3ca57fe9c975753bb0126589051908, I moved the `skip` step to the `Validate` method, therefore it would not return an error if we are asked to skip validation. However, because the `TransitiveReduction` should not run if there are validation errors, we have to do an additional early return in the transformer.

Reply:
Oh, I forgot that the transformer call drops the error! In that case yes, it is probably easiest to just return early here. Any errors that arise with the subgraphs are entirelyTerraform's problem, so not something a user would want to try debug with the graph.